### ISSUE TITLE
jfs_open may not support read-write mode or it's a bug in read-write mode

### ISSUE THREAD

**Issue by HugoAhoy** [juicedata/juicefs#655]
**What happened**:

Function `int jfs_open(long pid, long h, String path, int flags)` can't handle read-write mode(`os.R_OK | os.W_OK`). When trying to open a file with `flags = os.R_OK | os.W_OK`, we can only read the file use the file descriptor, but cannot write any content into the file.

I am not sure if it originally doesn't support read-write mode or it's a bug.

**What you expected to happen**:

After getting the file descriptor `fd`by `jfs_open` with `flags = os.R_OK | os.W_OK`, I supposed that both `jfs_pread` and `jfs_write` can work properly. In other words, it behaves exactly like `os` does.
But in out code snippet, we located that `jfs_write` doesn't properly.

**How to reproduce it (as minimally and precisely as possible)**:
Here is the python code
```python
def test_read_write_abp(jfs, filename):
    TEXT = b'hello world'
    jfs.create(filename, 0o777)
    fdw = jfs.open(filename, os.O_WRONLY)
    jfs.write(fdw, TEXT)
    jfs.close(fdw)

    fd = jfs.open(filename, os.O_RDWR)
    assert jfs.read(fd, 6) == TEXT[:6]
    jfs.write(fd, TEXT[::-1])
    jfs.flush(fd)

    jfs.lseek(fd, 0, os.SEEK_SET)

    # here raises AssertionError: assert b'hello world' == b'hello dlrow olleh'
    assert jfs.read(fd, 2*len(TEXT)) == TEXT[:6] + TEXT[::-1]
```

Here is the pseudo code.
```python
def test_read_write_abp():
    TEXT = b'hello world'
    TEXT_inv = TEXT[::-1]
    b = get_pointer(TEXT)
    b_inv = get_pointer(TEXT_inv)
    length = len(TEXT)

    # create a file with TEXT
    jfs_create(pid, h, filename, 0o777)
    fdw = jfs_open(pid, h, filename, os.W_OK)
    jfs_write(pid, fdw, b, length)
    jfs_close(pid, fdw)

    fd = jfs_open(pid, h, filename, os.R_OK | os.W_OK)
    
    # the 3 lines below correspond to jfs.read(fd, 6) in python code
    offset = jfs_lseek(pid, fd, 0, os.SEEK_CUR)
    content = jfs_pread(pid, fd, 6, offset)
    jfs_lseek(pid, fd, len(content), os.SEEK_CUR)

    assert content == TEXT[:6]

    jfs_write(pid, fd, b_inv, length)

    jfs_lseek(pid, fd, 0, os.SEEK_SET)
    offset = jfs_lseek(pid, fd, 0, os.SEEK_CUR)
    content = jfs_pread(pid, fd, 2*len(TEXT), offset)
    jfs_lseek(pid, fd, len(content), os.SEEK_CUR)

    # here raises AssertionError: assert b'hello world' == b'hello dlrow olleh'
    assert content == TEXT[:6] + TEXT_inv
```

**Anything else we need to know?**

1. The code snippet is written in Python, but the function is an easy python wrap of functions in `libjfs.so`. Here may I explain it.  
`jfs.XXX(args...)`(XXX is the function name) wraps the `jfs_XXX(pid, handle, args...)` or `jfs_XXX(pid, fd, args...)`. 
For example, `jfs.write(fd: int, content: bytes)` actually convert the `content` to a `Pointer buf` and call `jfs_writer(pid, fd, buf)`.  

2. `jfs.read(fd, size)` is the wrap of 3 functions in `libjfs.so`, as below shows:

```python
jfs.read(fd, size):
    offset = jfs_lseek(pid, fd, 0, os.SEEK_CUR)
    content = jfs_pread(pid, fd, size, offset)
    jfs_lseek(pid, fd, len(content), os.SEEK_CUR)    
```

**Environment**:
- JuiceFS version (use `./juicefs --version`) or Hadoop Java SDK version:
```
juicefs version 0.12.1 (2021-04-15T08:18:25Z 7b4df23)
```
- Cloud provider or hardware configuration running JuiceFS:
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`):
```
Linux determined-pare 4.15.0-55.60 #4 SMP Mon Mar 30 12:42:28 HKT 2020 x86_64 x86_64 x86_64 GNU/Linux
```
- Object storage (cloud provider and region):
- Redis info (version, cloud provider managed or self maintained):
- Network connectivity (JuiceFS to Redis, JuiceFS to object storage):
- Others:

**Comment by davies** [comment]
@HugoAhoy This should be fixed in latest main branch by #647 , could you verify that?

Since Java API does not support mixed read and write, so it will not affect existing workload.

**Comment by HugoAhoy** [comment]
> @HugoAhoy This should be fixed in latest main branch by #647 , could you verify that?
> 
> Since Java API does not support mixed read and write, so it will not affect existing workload.

@davies Based on our test, the read-write mode works properly on the main branch. Its behavior is exactly as Python os module does.

I will close this issue then. Thanks for your quick reply!
